File: Orimath/ViewModels/MenuItemViewModel.cs
using System.Collections.ObjectModel;
using System.IO;
using System.Windows.Input;
using Orimath.Plugins;
using Orimath.Controls;

namespace Orimath.ViewModels
{
    public class MenuItemViewModel : NotifyPropertyChanged
    {

        public string Name { get; }

        public KeyGesture? ShortcutKey { get; }

        public string ShortcutKeyText { get; }

        public ICommand? Command { get; }

        public Stream? IconStream { get; }

        public ObservableCollection<MenuItemViewModel> Children { get; } = new();

        public MenuItemViewModel(string name)
        {
            Name = name;
            ShortcutKeyText = "";
        }

        public MenuItemViewModel(string name, ICommand command)
        {
            Name = name;
            ShortcutKeyText = "";
            Command = command;
        }

        public MenuItemViewModel(IEffect effect, IMessenger messenger)
        {
            Name = effect.Name;
            Command = messenger.GetEffectCommand(effect);
            IconStream = effect.Icon;
            ShortcutKey = Internal.ConvertToKeyGesture(effect.ShortcutKey);
            ShortcutKeyText = ShortcutKey is not null ? effect.ShortcutKey : "";
        }
    }
}

File: Orimath/ViewModels/MenuItemViewModel.fs
namespace Orimath.ViewModels
open System.Collections.ObjectModel
open System.IO
open System.Windows.Input
open Orimath
open Orimath.Plugins
open Orimath.Controls

type MenuItemViewModel
    (name: string,
     shortcutKey: KeyGesture,
     shortcutKeyText: string,
     command: ICommand,
     iconStream: Stream
    ) =
    inherit NotifyPropertyChanged()
    
    member val Children = ObservableCollection<MenuItemViewModel>()
    member _.Name = name
    member _.ShortcutKey = shortcutKey
    member _.ShortcutKeyText = shortcutKeyText
    member _.Command = command
    member _.IconStream = iconStream

    new(name: string) = MenuItemViewModel(name, null, "", null, null)
    
    new(name: string, command: ICommand) = MenuItemViewModel(name, null, "", command, null)

    new(effect: IEffect, messenger: IMessenger) =
        let shortcutKey, shortcutKeyText =
            match Internal.convertToKeyGesture effect.ShortcutKey with
            | Some(sk) -> sk, effect.ShortcutKey
            | None -> null, ""
        MenuItemViewModel(effect.Name, shortcutKey, shortcutKeyText,
            messenger.GetEffectCommand(effect), effect.Icon)
